### PR TITLE
[Woptim#24] update spack, uberenv, radiuss spack configs and use spack user_cache_path

### PR DESCRIPTION
**Pull request by adrienbernede**
# Summary

- This PR is an update of CI parameters (Spack, Uberenv, RADIUSS-Spack-Configs)

# Design review

At the basis, the changes in this branch are the same as in task/add-tioga-ci-pipeline ( #1272 ), without the tioga specifics.
Thanks @kab163 for doing a good part of the update jobs.

Then I added changes to make sure Spack is used in an isolated environment, especially regarding clingo install. Those changes are based on @trws suggestions found here: https://github.com/spack/spack/issues/31030#issuecomment-1150085136.

I tested those change using may CI account. With this PR, the same test should run as @davidbeckingsale and give the same results, because clingo setup is now independent of the user config.

## CI results:

- [ ] There is one job failing, which should be looked at and fixed by RAJA team.
- [x] After comparing with Umpire, there is one last change needed, things were going too well...

## Misc.

- RAJA is not using a service account as Umpire does.
- More details in https://lc.llnl.gov/gitlab/radiuss/woptim/-/issues/24

### CODE REVIEW

**Comment by rhornung67** [comment]
@adrienbernede I don't see how your changes cause the CMake issue related to desul atomics? Do you have any ideas?

**Comment by rhornung67** [comment]
@adrienbernede @davidbeckingsale After other PRs were merged, I merged develop into this. Now, it appears that all the target export stuff is broken. I don't know what's going on here.

**Comment by adrienbernede** [comment]
David thought it could be because of a hidden BLT update.
The initial message suggests that the cuz target is defined in two different export sets. It would work if it were in only one.
I can try to pin point BLT version, it will have to wait Thursday.

**Comment by rhornung67** [comment]
@adrienbernede that's OK. We'll track it down.

**Comment by trws** [comment]
@adrienbernede, @rhornung67 and I poked at this some, and I just pushed what seems to be a minimal fix.  The short version is that the RAJA export set didn't contain the exported BLT targets, and the targets were getting exported twice somehow.  I'm not sure this will solve the whole problem, but hopefully one step closer.

**Comment by davidbeckingsale** [comment]
> @trws asking @davidbeckingsale to chime in. He says your change will break other things.

I'm second guessing myself now - let's see what the CI says. My  concern was the fact that raja-config.cmake looks for camp. Both depend on the BLT targets, but if they are in the RAJA export set, they won't be there when `find_dependency(camp)` is called.

**Comment by trws** [comment]
They do, camp for some reason generates a `camp::openmp` exported target despite not actually having a line of code requesting it (still not sure how/why) and it depends on that.  RAJA depends on the blt target, and needs to have it exported in the same export set, the error about multiple exports of the same target still has me worried, but I'm hoping this will get us closer.

**Comment by trws** [comment]
I should say, the thing I'm not sure about is the downstream impact, you reworked the export setup for  a reason as I recall @davidbeckingsale, and I certainly don't want to break app builds or something where they need the BLT targets as well.

**Comment by davidbeckingsale** [comment]
The export split fixed finding a camp that was built as part of RAJA - since it could depend on the BLT targets which were not imported until the RAJA targets were pulled in. But then adding target export to camp perhaps made the fix redundant because  camp now puts the targets in its own export set too.

**Comment by rhornung67** [comment]
When we think we have this stuff straightened out and we have tioga CI working, we should have a couple of apps try it out and then do a new coordinated release of RAJA, Umpire, camp, etc.

**Comment by rhornung67** [comment]
@davidbeckingsale something's amiss with the new cmake test you added. That appears to be what is breaking gitlab CI.

**Comment by davidbeckingsale** [comment]
Yup, that's what I was thinking would happen! But it's not the test that's wrong :p

**Comment by trws** [comment]
As far as I can tell with the tests I have on hand, this should export all the targets we need for both in-tree and out-of-tree camp builds.  The only visible difference should be that the `RAJA::<backend>` targets may or may not be named that, and there's no separate `raja-blt-targets.cmake` file because the necessary targets are part of the RAJATargets.cmake file. Both of those are meant to be an implementation detail, so I'm not sure how much we care about that, it's possible someone relies on those details.

**Comment by rhornung67** [comment]
@davidbeckingsale should we run the install test code in the CI as well? https://github.com/LLNL/RAJA/blob/develop/scripts/gitlab/build_and_test.sh#L200

**Comment by rhornung67** [comment]
@trws it looks like only the CUDA builds are failing now for the install test

**Comment by trws** [comment]
I'm not sure what's causing that one, doing a test with cuda, hopefully I missed switching a line like with cub last night, but the error in the CI output is pretty opaque, will let you all know what I find out when the test is done.

**Comment by rhornung67** [comment]
@trws I may be mistaken, but it looks to me that something's wrong with the link line for the install test when CUDA is enabled (i.e., can't find -lcuda_runtime).

**Comment by rhornung67** [comment]
@davidbeckingsale I meant add a line to the build_and_test.sh script to run the using-with-cmake executable after it is built, also add a print statement to the program at the end to indicate that it was run.

**Comment by trws** [comment]
That was definitely part of it, I managed to repro on ansel and updated camp gets farther, but it dies with this:

```
using-with-cmake /dev/shm/scogland1/install/lib/cmake/raja/RAJATargets.cmake Some (but not all) targets in this export set were already defined. CMake Error at /dev/shm/scogland1/install/lib/cmake/raja/RAJATargets.cmake:37 (message):
  Some (but not all) targets in this export set were already defined.

  Targets Defined: camp::cuda;camp::cuda_runtime

  Targets not yet defined: RAJA

Call Stack (most recent call first):
  /dev/shm/scogland1/install/lib/cmake/raja/raja-config.cmake:80 (include)
  CMakeLists.txt:16 (find_package)
```

We're back to the same exported targets from two components problem despite them being in completely separate export sets.

**Comment by trws** [comment]
I think I finally managed the exports.  The updates in here to make everything pass are not my favorite, but they work to be getting on with AFAICT.  This should almost certainly be propagated into BLT, with updates to all relevant places.  We also need a better solution for testing combined updates between camp and raja than the current uberenv setup.  The end solution is a lot like what @davidbeckingsale started to do with a secondary file for exported targets, but with some add-ons:

1. inject a fake target so the export set always has something in it
2. only export if the targets have not already been exported by someone else
3. fix propagation of OpenMP requirement to camp

I think the first two of these should probably become an extension to the way the export macro from BLT works, but wanted to get this working and tested before we reach out to yet another repo.

The gitlab failure is due to an issue that looks a whole lot like the issue Cyrus was having.  I think it has to do with not getting the right compiler propagated in, the ROCM_PATH thing is a red-herring because spack ensures everything is in the CMAKE_PREFIX_PATH, which renders ROCM_PATH redundant.  The error it's printing is what happens when /opt/rocm/bin/amdclang++ is selected instead of llvm/amdclang++ or similar, and we don't seem to be controlling the compiler selection at all so it may well be that.

There might also be an icpc error related to tuple, not sure what the deal is there.

**Comment by davidbeckingsale** [comment]
Looks like we need to add ROCPrim externals to the spack configuration...

**Comment by trws** [comment]
Thanks @davidbeckingsale and @rhornung67!  It looks like it's aaaalmost there, I fixed the module fallback for 3.20 being missing on corona, hopefully this time 🤞

File: scripts/gitlab/build_and_test.sh
Comment (davidbeckingsale) [inline scripts/gitlab/build_and_test.sh:142]:
```suggestion
    module load cmake/3.20.2 || module load cmake/3.19.2 || module load cmake/3.21.1
```

**Comment by rhornung67** [comment]
@davidbeckingsale It looks like lassen Gitlab CI went belly up. Do we need to ping someone to restart runners, etc.?

**Comment by trws** [comment]
Lassen is on emergency downtime from a cooling failure, so we may be a bit stuck on that right now.

On the upside, the corona tests ran *and passed* 🥳 🎆 

With that, I feel reasonably comfortable going forward on the camp changes, then we can start updating this to pull from the new camp update so we can start unspooling these stacked changes.

**Comment by rhornung67** [comment]
I'm in a meeting right now. If the lassen issue persists, I can try to manually test on rzansel in  the morning.

**Comment by rhornung67** [comment]
Something strange: All test pipleines are green in Gitlab, but it reports back as failing. Any ideas?

**Comment by davidbeckingsale** [comment]
> Something strange: All test pipleines are green in Gitlab, but it reports back as failing. Any ideas?

Looks green to me. Might need to refresh the page.

**Comment by rhornung67** [comment]
D'oh! Yup. Green Now. Thanks @davidbeckingsale and @trws. Are you folks working through the PR merges in the various projects so things get put in in the correct order? In particular, we have to wait on other PRs before merging this, correct?

**Comment by davidbeckingsale** [comment]
Yup this still needs to have camp and radiuss-spack-configs bumped to main. The camp update is waiting on a PR Tom has open: https://github.com/LLNL/camp/pull/112

**Comment by trws** [comment]
Ok, I *think* this is ready to go finally.  It still has workarounds in it for the AMD issues, and refers explicitly to `camp@main` in the raja package embedded for uberenv.  Some of this can go away when spack/spack#31739 lands, others when a new release is cut for everything.

After this goes in, based on discussions we've had, I think we need to look at the following cleanup to avoid having this be quite so painful next time:
1. promote some things to blt
    1. sycl support
    2. target export with the protections we developed here so they can be inherited across projects
2. add support in either build_and_test.sh or through the uberenv stuff somewhere for pulling the git hash of the submodules used in raja and passing them as constraints on the spack spec.  For example `^camp/deadbeef` might be generated, so we don't have to manually hack this stuff to get CI to work.
3. update our cmake requirement to 3.20 (anything over 3.16 could work, 3.20 has things we really want, and is just shy of the one that breaks a lot of projects)

**Comment by davidbeckingsale** [comment]
@trws do we need the logic you added to camp to look for `blt::` prefixed targets here as well?

**Comment by trws** [comment]
Good question, the final version of the camp and raja updates I cooked up here don't actually do that, but if as part of the blt update we also unconditionally add the aliases, and protect the creation of them in blt from errors caused by them already existing, then we probably should. It would make it easier to make sure composed projects always get the appropriate targets, but right now we didn't need it so I didn't pop it in.

**Comment by davidbeckingsale** [comment]
Ah gotcha,  that makes sense.

**Review by davidbeckingsale — APPROVED** [review]

**Comment by trws** [comment]
And of course, now that lassen is back up and everything else works, corona's TCE update is breaking us. 😖 

More importantly, the icpc 18 build is broken, though 19 is fine.  Are we still intentionally supporting 18 at this point @rhornung67?

**Comment by rhornung67** [comment]
No. Let's trash Intel 18.  A few apps are still on Intel 19. I'm trying to persuade them to move to Intel 2022 if they want an intel compiler. We can remove Intel 19 when everyone is off that.

**Review by rhornung67 — APPROVED** [review]
Thanks for working through all this mess.

**Comment by trws** [comment]
Thanks @rhornung67, the new corona TCE is currently breaking us because the rocm 5.1.0 doesn't exist anymore on there.  Corresponding PR on radiuss-spack-configs is already up.

**Comment by adrienbernede** [comment]
Awesome work thank you so much !

**Comment by rhornung67** [comment]
> Thanks @rhornung67, the new corona TCE is currently breaking us because the rocm 5.1.0 doesn't exist anymore on there. Corresponding PR on radiuss-spack-configs is already up.

Why does LC insist on removing earlier versions of compilers so quickly. We need  them around so we can do comparisons when issues arise. I thought we mentioned this to them before.

**Comment by trws** [comment]
Tioga still has them all, and corona still has some other older versions. I'm guessing something unfortunate happened on corona that broke it or corrupted it or something.

**Comment by rhornung67** [comment]
Nothing like debugging CI when things happen that you can't control, huh?

**Comment by rhornung67** [comment]
I think that when this PR is merged, we can turn off corona (if we don't need it) and move to tioga. Correct?

**Comment by trws** [comment]
Yes, and we'll get more relevant test results that way too honestly. Though be aware that it will be flipping to flux in a few weeks, shortly after corona does.

**Comment by adrienbernede** [comment]
hmm, that’s good to know. I’ll be following the Flux AWS tutorial, hopefully I’ll still have enough time to move scripts from slurm to flux...

**Comment by adrienbernede** [comment]
Also, we are planning to use radiuss-shared-ci on raja, so I’ll have to implement a tioga pipeline there.
Probably a good idea to test first on Umpire, what do you think @davidbeckingsale ?
Do RAJA and Umpire service accounts have access to Tioga already ?

**Comment by trws** [comment]
> hmm, that’s good to know. I’ll be attending the Flux AWS tutorial, hopefully I’ll still have enough time to move scripts from slurm to flux...

On this I'm hoping it will be minimal work to get going, Ryan Day is maintaining a set of wrappers for flux commands so `srun` and `sbatch` and similar should keep working for common options.  Only more complex cases will likely need meaningful tweaking.

If you have access to rzalastor, there's already a 4-node flux partition on there that's ready for testing, not sure if the wrappers have made it on there or not though.

File: .gitlab/ruby-jobs.yml
Comment (rhornung67) [inline .gitlab/ruby-jobs.yml:41]: 👍

File: .gitlab/ruby-jobs.yml
Comment (rhornung67) [inline .gitlab/ruby-jobs.yml:42]: 👍

**Comment by rhornung67** [comment]
@adrienbernede RAJA has not switched to use a service account for Gitlab CI yet. It runs as @davidbeckingsale

**Comment by trws** [comment]
It's green, it's finally ready to go.

**Comment by rhornung67** [comment]
Merge this or wait for other PRs to be merged (camp, blt, radiuss-spack-configs, etc.) first?

**Comment by trws** [comment]
The others are all merged except spack, but on spack I'm seeing umpire builds failing, @davidbeckingsale?

**Pull request merged**
